### DISCUSSION TITLE
A way to not inline Merge parameters

### DISCUSSION THREAD

**Discussion by Demivan** [linq2db/linq2db#4582] [Q&A] [answered]
I'm using Merge for upsert functionality with custom checks.

Example code:

```csharp
db.Likes
	.Merge()
	.Using(() => new LikeRecord
	{
		ItemId = itemId,
		ProfileId = profileId
	})
	.OnTargetKey()
	.InsertWhenNotMatchedAnd( // ...
```

this generates:
```sql
MERGE INTO [dbo].[Like] [Target]
USING (VALUES
	(2832038,3609294)
) [Source]
(
	[ItemId],
	[ProfileId]
)
ON --...
```

Because parameters are inlined, this creates lots of execution plans in database.
Is there a way to make `ItemId` and `ProfileId` query parameters when using Merge?

**Comment by MaceWindu** [comment]
Currently not possible. We do it to avoid hitting parameter count limit. Probably we can do something here.

**Reply by Demivan** [reply]
I managed to hack around this limitation. I used this as `source`:

```csharp
var source = db.FromSql<LikeRecord>(
	"""
	SELECT [ItemId], [ProfileId] FROM
	(VALUES ({0}, {1})) [Params]
	([ItemId], [ProfileId])
	""",
	itemId,
	profileId);
```

Maybe it would be a good idea to add `Using(Expression<Func<TSource>> source)` for upserting values using merge? Don't know whether this pattern is used that often.

**Reply by Demivan** [reply, marked as answer]
For those who find this issue later, instead of using `db.FromSql`, `db.SelectQuery(expression)` can be used to create the query needed for `Merge`. It does not work nicely with linq2db v5, but works with v6. This is what I use with v5:

```csharp
public static IMergeableOn<TTarget, TSource> UsingOne<TTarget, TSource>(
	this IMergeableUsing<TTarget> merge,
	DatabaseContext db,
	Expression<Func<TSource>> sourceExpression)
	where TSource : class
{
	var query = db.SelectQuery(sourceExpression)
		.Concat(Enumerable.Empty<TSource>()); // TODO: Remove concat when updating to linq2db 6.0.0
	return merge.Using(query);
}
```
